logstash-core-event/spec/logstash/timestamp_spec.rb

```diff
@@ -106,4 +107,50 @@
       expect(subject.to_f).to eq(now.to_f)
     end
   end
+
+  context "at" do
+    context "with integer epoch" do
+      it "should convert to correct date" do
+        expect(LogStash::Timestamp.at(946702800).to_iso8601).to eq("2000-01-01T05:00:00.000Z")
+      end
+
+      it "should return zero usec" do
+        expect(LogStash::Timestamp.at(946702800).usec).to eq(0)
+      end
+    end
+
+    context "with float epoch" do
+      it "should convert to correct date" do
+        expect(LogStash::Timestamp.at(946702800.123456.to_f).to_iso8601).to eq("2000-01-01T05:00:00.123Z")
+      end
+
+      it "should return usec with a minimum of millisec precision" do
+        expect(LogStash::Timestamp.at(946702800.123456.to_f).usec).to be_within(1000).of(123456)
+      end
+    end
+
+    context "with BigDecimal epoch" do
+      it "should convert to correct date" do
+        expect(LogStash::Timestamp.at(BigDecimal.new("946702800.123456")).to_iso8601).to eq("2000-01-01T05:00:00.123Z")
+      end
+
+      it "should return usec with a minimum of millisec precision" do
+        # since Java Timestamp relies on JodaTime which supports only milliseconds precision
+        # the usec method will only be precise up to milliseconds.
+        expect(LogStash::Timestamp.at(BigDecimal.new("946702800.123456")).usec).to be_within(1000).of(123456)
+      end
+    end
+  end
+
+  context "usec" do
+    it "should support millisecond precision" do
+      expect(LogStash::Timestamp.at(946702800.123).usec).to eq(123000)
+    end
+
+    it "should try to preserve and report microseconds precision if possible" do
+      # since Java Timestamp relies on JodaTime which supports only milliseconds precision
+      # the usec method will only be precise up to milliseconds.
+      expect(LogStash::Timestamp.at(946702800.123456).usec).to be_within(1000).of(123456)
```

Comment:
I would not change the current specs but also add randomized. let's defer that + create followup issue ok?